R/subscript-tables.R

```diff
@@ -0,0 +1,118 @@
+#' @export
+`[.QTable` <- function(x, ..., drop = TRUE) {
+    # Use sys.call as match.call captures unmatched named arguments into ...
+    used.arguments <- names(sys.call())
+    input.name <- attr(x, "name")
+    if (!validArgumentNames(used.arguments, "drop"))
+        throwErrorOnlyNamed("drop", "[")
+    if ("drop" %in% used.arguments && !is.logical(drop))
+        stop("drop argument should be TRUE or FALSE")
+    called.args <- match.call(expand.dots = FALSE)
+    empty.ind <- providedArgumentEmpty(called.args, optional.arg = "drop")
+    # Catch empty input e.g. x[] or x[drop = TRUE/FALSE] (when ... is empty)
+    if (empty.ind) {
+        y <- NextMethod(x)
+        class(y) <- c("QTable", class(y))
+        return(y)
+    }
+    x.dim <- dim(x)
+    n.dim <- length(x.dim)
+    n.index.args <- nargs() - 1L - !missing(drop)
+    # Throw a nicer error if the indexing is not appropriate
+    if (n.index.args != 1 && n.dim != n.index.args)
+        throwErrorTableIndexInvalid(input.name, n.dim, n.index.args)
+    y <- NextMethod(x)
+    class(y) <- c("QTable", class(y))
+    # Update Attributes here
+    attr(y, "name") <- input.name
+    y
+}
+
+#' @export
+`[[.QTable` <- function(x, ..., exact = TRUE) {
+    # Use sys.call as match.call captures the quoted arguments as names
+    used.arguments <- names(sys.call())
+    input.name <- attr(x, "name")
+    if (!validArgumentNames(used.arguments, "exact"))
+        throwErrorOnlyNamed("exact", "[[")
+    if ("exact" %in% used.arguments && !is.logical(exact))
+        stop("exact argument should be TRUE or FALSE")
+    called.args <- match.call(expand.dots = FALSE)
+    empty.ind <- providedArgumentEmpty(called.args, optional.arg = "exact")
+    x.dim <- dim(x)
+    if (empty.ind)
+        throwErrorEmptyDoubleIndex(input.name, x.dim)
+    n.dim <- length(x.dim)
+    n.index.args <- nargs() - 1L - !missing(exact)
+    correct.n.args <- n.index.args == n.dim
+    inputs <- as.list(called.args[["..."]])
+    all.unit.length <- all(lengths(inputs) == 1L)
+    if (!(correct.n.args && all.unit.length))
+        throwErrorTableDoubleIndex(input.name, x.dim)
+    y <- NextMethod(x)
+    class(y) <- c("QTable", class(y))
+    # Update Attributes here
+    attr(y, "name") <- input.name
+    y
+}
+
+validArgumentNames <- function(arg.names, optional.arg = NULL) {
+    all(arg.names %in% c("", optional.arg))
+}
+
+providedArgumentEmpty <- function(called.args, optional.arg) {
+    named.args <- names(called.args)
+    named.args[3L] == optional.arg || isEmptyList(called.args[3L])
+}
+
+isEmptyList <- function(x) x == quote(as.pairlist(alist())())
+
+describeTableMsg <- function(x.name, n.dim) {
+    type <- if (n.dim == 2) "a matrix" else "an array"
```

Comment:
I think it'd be nicer to avoid referencing R and R classes in this error message and just describe the problem and fix. E.g. `paste0("The supplied subscripts are invalid for table ", x.name, ". Please supply the subscripts as follows...")`.

Reply:
Agree that would be a bit easier to read and avoids extra maintenance clutter. I'll update it.

Reply:
Updated in https://github.com/Displayr/verbs/pull/39/commits/2f0c7d83ac69ac53775b9a22dbb30cbeaf89338a. Should now produce standardized errors like the following.
```
> x.6.5.4.3.2[1, 2]
Error in throwErrorTableIndexInvalid(input.name, x.dim) :
  The supplied subscripts on the Table (table.6.5.4.3.2) are invalid. When using the [
 subscript, either reference values with integers (or strings), or provide references
for each dimension. For example, table.6.5.4.3.2 can be subscriptted with, table.6.5.4
.3.2[1:3] or table.6.5.4.3.2[1:3, 1:2, 1:2, 3, 2].
> x.6.5.4.3.2[[1, 2]]
Error in throwErrorTableDoubleIndex(input.name, x.dim) :
  The supplied subscripts on the Table (table.6.5.4.3.2) are invalid. When using the [
[ subscript, there either needs to be a single integer (or string) reference, or one f
or each dimension. For example, table.6.5.4.3.2 can be subscriptted with, table.6.5.4.
3.2[[3]] or table.6.5.4.3.2[[3, 2, 2, 1, 1]].
```